daml-lf/scenario-interpreter/src/main/scala/com/digitalasset/daml/lf/speedy/ScenarioRunner.scala

```diff
@@ -296,26 +297,36 @@ object ScenarioRunner {
           missingWith(ScenarioErrorContractNotVisible(coid, tid, actAs, readAs, observers))
       }
     }
+
     override def lookupKey(
+        machine: Speedy.Machine,
         gk: GlobalKey,
         actAs: Set[Party],
         readAs: Set[Party],
-        canContinue: Option[ContractId] => Boolean,
+        callback: Option[ContractId] => Unit,
     ): Either[SError, Unit] =
-      handleUnsafe(lookupKeyUnsafe(gk, actAs, readAs, canContinue))
+      handleUnsafe(lookupKeyUnsafe(machine: Speedy.Machine, gk, actAs, readAs, callback))
 
     private def lookupKeyUnsafe(
+        machine: Speedy.Machine,
         gk: GlobalKey,
         actAs: Set[Party],
         readAs: Set[Party],
-        canContinue: Option[ContractId] => Boolean,
+        callback: Option[ContractId] => Unit,
     ): Unit = {
+
       val effectiveAt = ledger.currentTime
       val readers = actAs union readAs
 
-      def missingWith(err: SError) =
-        if (!canContinue(None))
-          throw SRunnerException(err)
+      def missingWith(err: SError) = {
+        callback(None)
+        machine.ctrl match {
+          case _: SExpr.SEDamlException =>
```

Comment:
I don’t understand this. Why can we not just continue interpretation and let it blow up?

Reply:
We want (do we really ?) rewrite the error in case of 
- `ScenarioLedger.LookupContractNotFound`
- `ScenarioLedger.LookupContractNotEffective`
- `ScenarioLedger.LookupContractNotActive`
To give a more precise error message. 

Should I kill `ScenarioLedger.LookupContractNotEffective` and `ScenarioLedger.LookupContractNotActive` ?

Reply:
oh I got confused, I thought we rethrow the `SEDamlException` but we rethrow what we pass to `missingWith`. I’m not entirely sure this is an improvement over what we had before. Now we depend on the implementation details of the callback here before instead of making it explicit by returning a boolean. But I don’t feel super strongly about this so if you prefer it fine to keep it.

Reply:
I do think we should keep it.